tests/library/chromium/chromium.spec.ts

```diff
@@ -250,6 +250,99 @@ test('setOffline', async ({ context, page, server }) => {
   expect(error).toMatch(/REJECTED.*Failed to fetch/);
 });
 
+test('should emit page-level request event for respondWith', async ({ page, server }) => {
+  await page.goto(server.PREFIX + '/serviceworkers/fetchdummy/sw.html');
+  await page.evaluate(() => window['activationPromise']);
+
+  // Sanity check.
+  const [pageReq, swResponse] = await Promise.all([
+    page.waitForEvent('request'),
+    page.evaluate(() => window['fetchDummy']('foo')),
+  ]);
+  expect(swResponse).toBe('responseFromServiceWorker:foo');
+  expect(pageReq.url()).toMatch(/fetchdummy\/foo$/);
+  expect(pageReq.serviceWorker()).toBe(null);
+  expect((await pageReq.response()).fromServiceWorker()).toBe(true);
+});
+
+test('should emit page-level request event for respondWith for cached resource', async ({ context, page, server }) => {
+  test.fixme();
+  test.info().annotations.push({ type: 'issue', description: 'https://github.com/microsoft/playwright/issues/15474' });
+
+  await context.route(/service-worker\.js/, route => route.fulfill({
+    status: 200,
+    contentType: 'application/javascript',
+    body: `
+        self.addEventListener("activate", event => {
+          event.waitUntil(clients.claim());
+        });
+        self.addEventListener("fetch", async (event) => {
+          event.respondWith((async () => {
+            const cachedResponse = await caches.match(event.request);
+            if (cachedResponse) {
+                return cachedResponse;
+            }
+            return new Response(JSON.stringify({ message: 'Hello from service-worker!' }), { status: 200, headers: { 'content-type': 'application/json' } });
+          })());
+        });
+          `
+  }));
+  await context.route(/\/test-page.html/, route => route.fulfill({
+    status: 200,
+    contentType: 'text/html',
+    body: `
+        <!DOCTYPE html>
+        <html>
+          <body>
+            <script>
+              window.serviceWorkerPromise = navigator.serviceWorker.register('/service-worker.js');
+              window.activationPromise = navigator.serviceWorker.ready;
+            </script>
+          </body>
+        </html>
+      `
+  }));
+  await context.route(/scripts\/resource-1.js/, (route, request) => {
+    if (request.serviceWorker()) {
+      route.fulfill({
+        status: 200,
+        contentType: 'application/javascript',
+        body: 'window.PW_VALUE_FROM_LOADED_RESOURCE = true;',
+      });
+    } else {
+      route.continue();
+    }
+  });
+  const [ sw ] = await Promise.all([
+    context.waitForEvent('serviceworker'),
+    page.goto(server.PREFIX + '/test-page.html'),
+    page.waitForFunction('activationPromise'),
+  ]);
+
+  await Promise.all([
+    context.waitForEvent('request', request => !!request.serviceWorker()),
+    sw.evaluate(async () => {
```

Comment:
Perhaps, just reloading the page would trigger the cache response?

Reply:
Good point—let me see if it repros like that instead.

Reply:
Oh—we don't load this script on page-load, so this is the first time we're requesting it.